### PR TITLE
rubocops/cask: Enforce the use of `on_{system}` blocks

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- Sort of related to #14898.
- As discussed in https://github.com/Homebrew/brew/pull/14976#issuecomment-1474544569 and further comments, this is needed because in order to enforce the order of `on_{arch,system}` blocks we need to have everything consistently within one of those blocks.
- We previously allowed overrides where the top-level `version` stanza would be the default, unless on an OS that had an `on_system` block with a `version` specified. But this breaks down when we try to order the `on_system` blocks because if a `url` at the top-level has a `version` interpolated in it, then the `version` stanza needs to be above the `url` stanza. But it could be that `version` is OS-specific.
- Let's stop allowing overrides and require that every stanza already in an `on_system` block be in a companion `on_system` block. This will make it easier to enforce the order of `on_system` blocks in a future PR (#14976).

### CODE REVIEW

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        EOS
+
+        def on_cask(cask_block)
+          return if cask_block.toplevel_stanzas.empty?
+
+          cask_block.toplevel_stanzas.each do |stanza|
+            # TODO: We probably only want to disallow `version`, `url`, and `sha256` stanzas being overridden?
+            next unless RuboCop::Cask::Constants::STANZA_ORDER.include?(stanza.stanza_name)
+            # Skip if the stanza we detect is already in an `on_*` block.
+            next if stanza.parent_node.block_type? && stanza.parent_node.method_name.to_s.start_with?("on_")
```
Comment (Rylan12) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:23]:
```suggestion
            next if stanza.parent_node.block_type? && RuboCop::Cask::Constants::ON_SYSTEM_METHODS.include?(stanza.parent_node.method_name)
```

**Review by Rylan12 — COMMENTED** [review]
Does this work for cases with no `on_{system}` blocks at all? We only want to force an `on_{system}` block if needed. This kind of thing should be fine:

```ruby
cask "foo" do
  version "1.2.3"
  sha256 "aaa"
  
  url "https://brew.sh/foo-2.3.4.dmg"
end
```

File: Library/Homebrew/rubocops/cask/no_overrides.rb (hunk 1 above)
Comment (issyl0) [reply Library/Homebrew/rubocops/cask/no_overrides.rb:23]: Blame Copilot for that one. 😉

**Comment by issyl0** [comment]
> Does this work for cases with no on_{system} blocks at all?

Now it does.

> And in cases where the only difference is e.g. the url, we also don't want to force it for e.g. version, right?

I can't get my head around this right now. Hopeful for a 💡🧠 soon!

**Comment by reitermarkus** [comment]
> I can't get my head around this right now. Hopeful for a 💡🧠 soon!

We need to first collect all stanzas which are present in `on_*` blocks, then check if those are also present at the top level.

So for

```ruby
cask "foo" do
  version "1.2.3"
  sha256 "aaa"
  url "https://brew.sh/foo-2.3.4.dmg"

  on_big_sur :or_older do
    sha256 "bbb"
    url "https://brew.sh/legacy/foo-2.3.4.dmg"
  end
end
```

we collect the set `[:sha256, :url]` from the `on_*` blocks, so we need to flag the top-level `sha256` and `url` stanzas.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by issyl0** [comment]
OK, running against actual Casks, this reports `4184 files inspected, 89 offenses detected`.

**Comment by issyl0** [comment]
I’ll do a PR this week to fix up the Cask failures manually and see how bad/confusing it is.

It might be a bit tedious to go through and correct all of those. Maybe it would be better for users if we could somehow specify what the `on_*` block syntax they should use is, particularly the macOS version? But that’s almost an autocorrect then and that seems complicated.

File: Library/Homebrew/test/support/fixtures/cask/Casks/multiple-versions.rb [hunk 2]
```diff
@@ -2,9 +2,10 @@
   arch arm: "arm", intel: "intel"
   platform = on_arch_conditional arm: "darwin-arm64", intel: "darwin"
 
-  version "1.2.3"
-  sha256 "67cdb8a02803ef37fdbf7e0be205863172e41a561ca446cd84f0d7ab35a99d94"
-
+  on_monterey :or_newer do
+    version "1.2.3"
+    sha256 "67cdb8a02803ef37fdbf7e0be205863172e41a561ca446cd84f0d7ab35a99d94"
+  end
```
Comment (reitermarkus) [inline Library/Homebrew/test/support/fixtures/cask/Casks/multiple-versions.rb:8]: The order of these should be reversed, oldest to newest.

**Review by Rylan12 — COMMENTED** [review]
This looks good to me. Just one edge-case question: does this handle nested blocks like this:

```ruby
cask "foo" do
  version "1.2.3"

  on_intel do
    on_monterey :or_older do
      version "1.0.0"
    end
    on_ventura :or_newer do
      version "1.1.0"
    end
  end
end
```

It looks like `on_blocks` only contains the _top-level_ `on_{system}` blocks, so I'm thinking that `on_system_stanzas` will never see the `version` stanzas

File: Library/Homebrew/test/support/fixtures/cask/Casks/multiple-versions.rb (hunk 2 above)
Comment (issyl0) [reply Library/Homebrew/test/support/fixtures/cask/Casks/multiple-versions.rb:8]: Right, yes, that was the other PR's change to enforce that ordering. But the ordering makes sense to fix here from an example flow perspective. Lots of things to keep up with. 😅

**Comment by reitermarkus** [comment]
In this case the top-level `version` should be in an `on_arm` block though, so this cop wouldn't catch it.

How would we handle 

```ruby
cask "foo" do
  version "1.2.3"

  on_intel do
    on_monterey :or_older do
      version "1.0.0"
    end
  end
end
```

though? Duplicate `version` in `on_arm` and in `on_intel { on_ventura :or_newer }`?

**Comment by reitermarkus** [comment]
I guess my example is more likely this

```ruby
cask "foo" do
  on_monterey :or_older do
    version "1.0.0"

    depends_on arch: :intel
  end
  on_ventura :or_newer do
    version "1.2.3"
  end
end
```

**Comment by reitermarkus** [comment]
I'd say let's forget about the edge case for now, since when mixing `on_arch` and `on_system` we would also have to take `depends_on arch:` into account.

**Comment by Rylan12** [comment]
Yeah, I mean that's an edge case that is definitely worse-looking when we don't allow the overrides.

**Comment by issyl0** [comment]
I am so confused by these edge cases and what we're tracking or not. 😆

**Comment by issyl0** [comment]
> There's either a 🐛 or an oversight or this behaviour for arch is actually intentional?

Heh, it is a 🐛, the "stanza names" array looks like `[:version, :sha256, :url, :no_dots, :version, :version, :sha256, :url, :arch]` for that cask. But `arch` is a variable. Some more work to do here!

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        include CaskHelp
+
+        ON_SYSTEM_METHODS = RuboCop::Cask::Constants::ON_SYSTEM_METHODS
+        MESSAGE = <<~EOS
+          Do not use a top-level `%<stanza>s` stanza as the default. Add it to an `on_{system}` block instead.
+          Use `:or_older` or `:or_newer` to specify a range of macOS versions.
+        EOS
+
+        def on_cask(cask_block)
+          return if (cask_stanzas = cask_block.toplevel_stanzas).empty?
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:18]: No need for this check since looping makes an empty array a no-op anyways.

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+          # Skip if there are no `on_*` blocks.
+          return unless (on_blocks = cask_stanzas.select { |s| ON_SYSTEM_METHODS.include?(s.stanza_name) }).any?
+
+          cask_stanzas.each do |stanza|
+            # Skip if the stanza is itself an `on_*` block.
+            next if ON_SYSTEM_METHODS.include?(stanza.stanza_name)
+            # Skip if the stanza we detect is already in an `on_*` block.
+            next if stanza.parent_node.block_type? && ON_SYSTEM_METHODS.include?(stanza.parent_node.method_name)
+            # Skip if the stanza outside of a block is not also in an `on_*` block.
+            next if on_system_stanzas(on_blocks).none?(stanza.stanza_name)
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:28]: Call `on_system_stanzas(on_blocks)` before the loop.

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        def on_cask(cask_block)
+          return if (cask_stanzas = cask_block.toplevel_stanzas).empty?
+          # Skip if there are no `on_*` blocks.
+          return unless (on_blocks = cask_stanzas.select { |s| ON_SYSTEM_METHODS.include?(s.stanza_name) }).any?
+
+          cask_stanzas.each do |stanza|
+            # Skip if the stanza is itself an `on_*` block.
+            next if ON_SYSTEM_METHODS.include?(stanza.stanza_name)
+            # Skip if the stanza we detect is already in an `on_*` block.
+            next if stanza.parent_node.block_type? && ON_SYSTEM_METHODS.include?(stanza.parent_node.method_name)
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:26]: Do we need this check? Doesn't `cask_block.toplevel_stanzas` return only the top-level stanzas?

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            # Skip if the stanza outside of a block is not also in an `on_*` block.
+            next if on_system_stanzas(on_blocks).none?(stanza.stanza_name)
+
+            add_offense(stanza.source_range, message: format(MESSAGE, stanza: stanza.stanza_name))
+          end
+        end
+
+        def on_system_stanzas(on_system)
+          names = []
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:35]:
```suggestion
          names = Set.new
```

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+          method_nodes = on_system.map(&:method_node)
+          method_nodes.each do |node|
+            next unless node.block_type?
+
+            node.child_nodes.each do |child|
+              child.each_node(:send) do |send_node|
+                next if ON_SYSTEM_METHODS.include?(send_node.method_name)
+
+                names << send_node.method_name
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:44]:
```suggestion
                names.add(send_node.method_name)
```

File: Library/Homebrew/rubocops/cask/no_overrides.rb (hunk 6 above)
Comment (issyl0) [reply Library/Homebrew/rubocops/cask/no_overrides.rb:36]: Yeah I was gonna do this, `names` gets unwieldy otherwise.

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+          # Skip if there are no `on_*` blocks.
+          return unless (on_blocks = cask_stanzas.select { |s| ON_SYSTEM_METHODS.include?(s.stanza_name) }).any?
+
+          cask_stanzas.each do |stanza|
+            # Skip if the stanza is itself an `on_*` block.
+            next if ON_SYSTEM_METHODS.include?(stanza.stanza_name)
+            # Skip if the stanza we detect is already in an `on_*` block.
+            next if stanza.parent_node.block_type? && ON_SYSTEM_METHODS.include?(stanza.parent_node.method_name)
+            # Skip if the stanza outside of a block is not also in an `on_*` block.
+            next if on_system_stanzas(on_blocks).none?(stanza.stanza_name)
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:29]:
```suggestion
            next unless on_system_stanzas(on_blocks).include?(stanza.stanza_name)
```

File: Library/Homebrew/rubocops/cask/no_overrides.rb (hunk 5 above)
Comment (issyl0) [reply Library/Homebrew/rubocops/cask/no_overrides.rb:29]: We do need this because `toplevel_stanzas` includes `version` and it's fine if the `version` stanza in the Cask is in an `on_*` block already.

(To be sure, I removed this `next` and the "RuboCop::Cop::Cask::NoOverrides when there are no top-level standalone stanzas does not report any offenses" test failed.)

**Comment by issyl0** [comment]
OK, with the bug fixes I've made, this detects in "" in Cask right now. I think this might be ready, pending a PR in Cask that I'll do at some point this week to show the example fixes. And we still haven't decided if we should attempt some kind of autocorrect? I have to say flipping between the Wikipedia page for macOS versions to get the codenames to put them in the right `:or_newer`, `:or_older` order is quite tedious. But we should only have to do this mass correction once?

**Comment by issyl0** [comment]
There's another bug here with some real casks, the handling of "skip stanzas not already in `on_*` blocks is bugged. For Cask `sip`, as a random example:

```ruby
  on_mojave :or_newer do
    version "2.7.1"
    sha256 "074c7a1b5e4ebfabdc43b6583a96da9dfaf02f1a93a468b616dc14833ea1705d"

    livecheck do
      url "https://sipapp.io/updates/"
      regex(%r{href=.*?/sip[._-]v?(\d+(?:\.\d+)*)\.dmg}i)
    end
  end

  url "https://sipapp.fra1.digitaloceanspaces.com/updates/v#{version.major}/sip-#{version}.dmg",
      verified: "sipapp.fra1.digitaloceanspaces.com/updates/"
```

Since `url` is in `on_mojave` the top-level `url` stanza detected as in the block even though it's a _livecheck_ URL not a "standard" URL for the download.

File: Library/Homebrew/rubocops/cask/no_overrides.rb (hunk 5 above)
Comment (reitermarkus) [reply Library/Homebrew/rubocops/cask/no_overrides.rb:29]: > We do need this because `toplevel_stanzas` includes `version` and it's fine if the `version` stanza in the Cask is in an `on_*` block already.

So `cask_block.toplevel_stanzas` includes `version` if the cask only contains the following?

```ruby
on_mojave :or_later do
  version :latest
end
```

If that is the case, that doesn't seem right.

**Comment by reitermarkus** [comment]
> Since `url` is in `on_mojave` the top-level `url` stanza is detected as in the block even though it's a _livecheck_ URL not a "standard" URL for the download.

Judging from this and https://github.com/Homebrew/brew/pull/15013/files#r1142791224, the `toplevel_stanzas` method seems to be bugged and returns more than just top-level stanzas. I haven't looked into whether that's actually the case yet though.

**Comment by reitermarkus** [comment]
https://github.com/Homebrew/brew/blob/ffa7ee01126c3ae59c80e684550702ccaa5aeba9/Library/Homebrew/rubocops/cask/ast/stanza.rb#L56

`toplevel_stanza?` seems to check two levels up for a `cask` block, no idea why.

File: Library/Homebrew/rubocops/cask/no_overrides.rb (hunk 5 above)
Comment (issyl0) [reply Library/Homebrew/rubocops/cask/no_overrides.rb:29]: This is because `toplevel_stanzas` is https://github.com/Homebrew/brew/blob/ffa7ee01126c3ae59c80e684550702ccaa5aeba9/Library/Homebrew/rubocops/cask/ast/cask_block.rb#L38 and that `toplevel_stanza?` is defined as https://github.com/Homebrew/brew/blob/ffa7ee01126c3ae59c80e684550702ccaa5aeba9/Library/Homebrew/rubocops/cask/ast/stanza.rb#L55-L57. So it _does_ support one level of nesting. That explains why I had to have this condition to satisfy that test.

**Comment by reitermarkus** [comment]
Okay, so I found one problem with `toplevel_stanzas` so far. For the following 

```ruby
cask do
  url
end
```

the AST looks like `cask > url`, so `cask` is the parent node of `url`.

For 

```ruby
cask do
  url
  homepage
end
```

the AST is actually `cask > block > url`, so `cask` is the grandparent node of `url`.

**Comment by reitermarkus** [comment]
> I don't understand how this can be?

If there is only one stanza, there is no need to group it.

In the AST it's similar to how an `if` statement works:

```ruby
stmt1 if cond
```

doesn't need a block, while

```ruby
if cond
  stmt1
  stmt2
end
```

does.

**Comment by issyl0** [comment]
Rebased on `master` after #15028.

**Comment by issyl0** [comment]
There's something funky now with the variable interpolation detection, but this is _very close_ to being done now hopefully!

File: Library/Homebrew/rubocops/cask/no_overrides.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+          names = Set.new
+          method_nodes = on_system.map(&:method_node)
+          method_nodes.each do |node|
+            next unless node.block_type?
+
+            node.child_nodes.each do |child|
+              child.each_node(:send) do |send_node|
+                # Skip (nested) livecheck blocks as its `url` is different to a download `url`.
+                next if send_node.method_name == :livecheck
+                next if send_node.parent.block_type? && send_node.parent.method_name == :livecheck
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:45]: I think we should have a helper function for this. This is the same check that is needed in `toplevel_stanzas`, but for `livecheck` instead of `cask`, right?

Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/no_overrides.rb:45]: I think we should have a helper function for this. This is the same check that is needed in `toplevel_stanzas`, but for `livecheck` instead of `cask`, right?

**Comment by issyl0** [comment]
OK, I think we're there (_n_th time lucky - it's got to be more than 3 by now)... 🙏🏻 The main Cask repo reports `4183 files inspected, 35 offenses detected` which is significantly less than there were in previous days when I found out the brokenness, so I think the main 🐛 are squashed.

**Comment by issyl0** [comment]
OK, there's a PR to fix the `Homebrew/homebrew-cask` problems at https://github.com/Homebrew/homebrew-cask/pull/143646. Thanks to that exercise I caught a bunch more edge cases! This is now, I think, ready for real review!

**Comment by reitermarkus** [comment]
As seen in https://github.com/Homebrew/homebrew-cask-versions/pull/16257, the cop needs to differentiate between stanzas which can be overridden (e.g. `url`) and those which cannot (e.g. `app`, `binary`).

**Comment by issyl0** [comment]
@reitermarkus Thanks, I'm learning lots about Casks! Is there a list somewhere of parts that are overridable and those that are not?

**Comment by reitermarkus** [comment]
I don't think we have a list, I think it's implemented for each stanza individually.

**Review by Rylan12 — APPROVED** [review]
Looks good to me once the homebrew/cask PR gets through

Also, if you haven't already, make sure to check for style issues in the other cask taps since there might be issues there and `brew style` usually runs on them in a later step. I always forget to check and then when everything in `homebrew/cask` is finally ready I rerun and realize there are a bunch more to fix 😅

**Review by meththot — COMMENTED** [review]
Ty
